### PR TITLE
Eliminate streamdevice from IMS IOC

### PR DESCRIPTION
**Pull request by mcb64**
Streamdevice and the imsRecord don't seem to play well together, so this PR will eliminate the use of streamdevice in IMS IOCs.

Currently, there are three ways that streamdevice is used:

- An I/O Intr stringin record reads the motor status and writes it to the .SSTR field of the motor.
- An I/O Intr stringin record reads the "Want2Save" message from the motor and writes "Save" to the .SVNG field of the motor.
- An I/O Intr stringin record reads the "Saved xxx" message from the motor and write "xxx" (the saved position) to the .SVNG field of the motor as well as to :SAVED_STR, which uses it to populate :SAVED_P (the saved position) and :SAVED_TS (the time of the save).

Now, a new thread, io_controller is used to do all asyn I/O.  A new routine has been written, send_recv_reply to combine the send and response into one routine.  All of the commands requiring responses being "PR", and the code has been changed to ensure that each printed response begins with two unique characters, allowing an easy way of matching what response belongs to which request.  The one exception is user-created commands (written to the .CMD field), but these are treated as a wildcard match.

io_controller handles the three streamdevice cases mentioned above internally.

### CODE REVIEW

File: imsApp/Db/ims_long.db
Comment (joshc-slac) [inline imsApp/Db/ims_long.db:678]: Nit: do we want this description to be slightly different than the DESC field of the $(MOTOR):SAVED_RAW record below

File: imsApp/Db/ims_long.db
Comment (joshc-slac) [inline imsApp/Db/ims_long.db:706]: Just for me an EPICS noob, I see how we have used the common record property FLNK to link together these trailing four records. I am assuming the point is to capture the functionality you mention in the PR:
>
    - An I/O Intr stringin record reads the motor status and writes it to the .SSTR field of the motor.
    - An I/O Intr stringin record reads the "Want2Save" message from the motor and writes "Save" to the .SVNG field of the motor.
    - An I/O Intr stringin record reads the "Saved xxx" message from the motor and write "xxx" (the saved position) to the .SVNG field of the motor as well as to :SAVED_STR, which uses it to populate :SAVED_P (the saved position) and :SAVED_TS (the time of the save).

If you could say a little more about what these "motor fields" are it would be helpful. My second assumption is that they tie into the Beckhoff motor control layer(NC)?

File: imsApp/Db/ims_short.db
Comment (joshc-slac) [inline imsApp/Db/ims_short.db:650]: Nit: unique DESC would be informative

File: imsApp/src/ims.h
Comment (joshc-slac) [inline imsApp/src/ims.h:20]: Again, exposing my EPICS naivete; I will attempt to build this project to verify but within this h file where do we define epicsEvent? I see the Makefile in this directory and that may connect the two; explicit includes are also nice. Apologies if I'm missing something obvious will know after building

Comment (joshc-slac) [reply imsApp/src/ims.h:20]: https://github.com/slac-epics/IMS/blob/5d3615fd2472b8c70cbf17e2deb008b9e0c93d4b/imsApp/src/imsRecord.cc#L34C16-L34C16

I would imagine this line is why it works, would mean this h file wouldn't work elsewhere

File: imsApp/src/imsRecord.cc
Comment (joshc-slac) [inline imsApp/src/imsRecord.cc:119]: Nit: not specific to this PR but the codebase. This file is far too long to be maintainable. This whole section and the above should go in its own h file minimally. (Not at all saying thats on @mcb64 just making a note)

File: imsApp/src/imsRecord.cc
Comment (joshc-slac) [inline imsApp/src/imsRecord.cc]: I am not sure if we added clarity by splitting this single line if into two lines. -Werror would tell you this is dangerous, I would add brackets if we want to improve

File: imsApp/src/imsRecord.cc
Comment (joshc-slac) [inline imsApp/src/imsRecord.cc:149]: Nit: especially because this is the *only* instance of malloc being used in this multi thousand line file, lets replace minimally with new but even better a smart pointer. This would be in a future PR

File: imsApp/src/imsRecord.cc
Comment (joshc-slac) [inline imsApp/src/imsRecord.cc:157]: I really would like out of bound access to be checked here. There's also a world where we use a foreach loop to guard against OOB access, not sure what C standard we are compiling to

File: imsApp/src/imsRecord.cc
Comment (joshc-slac) [inline imsApp/src/imsRecord.cc:169]: NIT: if a PR is cut to bring this codebase to the c11 standard address this too

File: imsApp/src/imsRecord.cc
Comment (joshc-slac) [inline imsApp/src/imsRecord.cc:3660]: Again not on you as this file is already 300 lines long, would be cool to put this functionality in its own file with accompanying header. In a more sane repository / library environment we could then share this functionality across repos and only make improvements in one place

File: imsApp/src/imsRecord.cc
Comment (joshc-slac) [inline imsApp/src/imsRecord.cc:3669]: I would vastly prefer a while(!killsignal) schema here. Not deal breaking but nice to be able to invoke cleaner kill signals when you try join detached threads (understand that is the off nominal case atm)

File: imsApp/src/imsRecord.cc
Comment (joshc-slac) [inline imsApp/src/imsRecord.cc:3670]: Nit: Could use the classic condition_variable with `.wait_until` to achieve the same "work notification" schema.  
However, I am not sure what updates the state member of this object. 

I am not sure what speed you want this thread to be running at but I believe this would be a better flow control mechanism (or at least more explicit)

File: imsApp/src/imsRecord.cc
Comment (joshc-slac) [inline imsApp/src/imsRecord.cc:3699]: many really cool logging libraries like spdlog, if we don't want to do debug this way. not for this PR of course

File: imsApp/src/imsRecord.cc
Comment (joshc-slac) [inline imsApp/src/imsRecord.cc]: Because this is new code, a slightly more informative comment would be appreaciative so future devs can take a look at this line when debugging something totally unrelated and know what you are getting across

File: imsApp/src/imsRecord.cc
Comment (joshc-slac) [inline imsApp/src/imsRecord.cc]: These two operations seem clever but dicey. For string parsing I really prefer the std::string implemention; there are often performance boosts as well not to mention the robusitification inherent in not dealing with raw pointers indexing into effectively arbitrary strings parsed of the line. 

(Again, if this PR is well tested and functional lets cut a future PR for the type of improvement I am suggesting in this comment. If this PR is still not _fully_ tested, it may be worth importing \<string\>)

File: imsApp/src/imsRecord.cc
Comment (joshc-slac) [inline imsApp/src/imsRecord.cc]: MAJOR NIT: the `auto` keyword is one of the coolest new features in since c11 (iirc). It makes things a lot more readable, the heuristic I made up for myself: "auto should be used when its obvious what type something should be and cleans up code readability; do not use auto just cause you're lazy". 

Apologies for so many nits :p

File: imsApp/src/imsRecord.cc
Comment (joshc-slac) [inline imsApp/src/imsRecord.cc]: OOB check before access of a member variable which itself is an array can be prudent

File: imsApp/src/imsRecord.cc
Comment (joshc-slac) [inline imsApp/src/imsRecord.cc:3535]: I like this paranoia! We could use a semaphore / condition variable do explicitly signal between threads

File: imsApp/src/imsRecord.cc
Comment (joshc-slac) [inline imsApp/src/imsRecord.cc:3537]: Really do not like this mutli line ifs without brackets (or proper indentation). Signal line is kinda fine; -Werror and -Wall would freak out about this

File: sim/Makefile
Comment (joshc-slac) [inline sim/Makefile:7]: MAJOR NIT clean is always appreaciated

.PHONY: clean
clean:
  rm server client

**Review by joshc-slac — COMMENTED** [review]
Short version: If tested and functional: LGTM.

Longer version: this repo has sins that predate this PR by 10 years. Slight refactors to align with the c11 industry standard are low hanging fruits that could lead to major robustification / maintainability improvements to this piece of software.

File: imsApp/src/imsRecord.cc
Comment (ZLLentz) [inline imsApp/src/imsRecord.cc:1]: I might need another pass through the diff here another day, but I didn't see any errors in the logic and each `lock` was accompanied by an `unlock`. I found the replacements for the on-record db file/proto file asyn handling and they look appropriate.

File: imsApp/Db/ims_long.db
Comment (ZLLentz) [inline imsApp/Db/ims_long.db:1]: I double-checked the screens and the python to confirm that all of the removed PVs are not used in any external capacity.

File: imsApp/src/imsRecord.cc
Comment (ZLLentz) [inline imsApp/src/imsRecord.cc:3530]: Why are we allowed to unlock here before waiting for the response? Is there no way for us to receive the wrong message if we send two messages at once from different threads? Or are we simply always processing these in such a way that the order doesn't matter?

**Review by ZLLentz — COMMENTED** [review]
I'd like to look again another day but I think I largely understand what you've implemented here and the functionality looks good. I don't want to hit approve until a few others have looked and after I've also looked again.

File: imsApp/src/imsRecord.cc
Comment (mcb64) [reply imsApp/src/imsRecord.cc:3530]: I am using rMutex to control access to the request array.  If I don't unlock, io_controller will never be able to process anything!

File: imsApp/Db/ims_long.db
Comment (mcb64) [reply imsApp/Db/ims_long.db:678]: They are the same value, different formats (one string, one int).  I don't think it's important to make this distinction, especially since we probably won't look at this field anyway.

File: imsApp/Db/ims_long.db
Comment (mcb64) [reply imsApp/Db/ims_long.db:706]: The imsRecord is a custom record that, frankly, has some pretty obscure functionality.  In particular:
- SSTR expects to get the last half of the status message (the status bits followed by the position).
- SVNG expects to get either "Save" when the motor should be ordered to save its state, or an integer in the form of a string, giving the position that was saved.

In other words, these fields are used to process the messages that the IMS program sends us, and nothing more or less.

File: imsApp/src/ims.h
Comment (mcb64) [reply imsApp/src/ims.h:20]: This is in one of the epics standard include files... probably epicsEvent.h, which is probably included by something else.  In practice, ims.h is only used in imsRecord.h and is included *after* the other includes.  Bad practice, I suppose, but there it is... I didn't really want to clean up after years of programming, just fix this little problem.

File: imsApp/src/imsRecord.cc
Comment (mcb64) [reply imsApp/src/imsRecord.cc]: I originally wanted to do something in the second pass, but it turned out to be unnecessary.  I accidentally didn't return the code back to the original form.

File: imsApp/src/imsRecord.cc
Comment (ZLLentz) [reply imsApp/src/imsRecord.cc:3530]: Ahh I see some of my misunderstanding now

File: imsApp/src/imsRecord.cc
Comment (mcb64) [reply imsApp/src/imsRecord.cc:157]: REQ_CNT is the constant used to dimension the array.  I'm not sure what more you want here, unless your comment belonged on a different line.

File: imsApp/src/imsRecord.cc
Comment (mcb64) [reply imsApp/src/imsRecord.cc:3669]: EPICS processes never quit, they die unexpectedly. :-)

File: imsApp/src/imsRecord.cc
Comment (mcb64) [reply imsApp/src/imsRecord.cc:3670]: The thread is gated by the timeout for the read.  I expect it to almost always be waiting for input, but when that times out, it checks if it has new work, and dispatches it immediately.

I think I set the timeout at 0.1s, so 10Hz?  Ish.

File: imsApp/src/imsRecord.cc
Comment (mcb64) [reply imsApp/src/imsRecord.cc]: I can clarify.  But before I edit the comment, I'm not sure that the read will *ever* give me two messages at once.  (Perhaps the kernel read does, but I tend to doubt that.  But the asyn read almost certainly checks for the EOS characters.)

Paranoia.

File: imsApp/src/imsRecord.cc
Comment (mcb64) [reply imsApp/src/imsRecord.cc]: Again, the size of the array is REQ_CNT.  So the for *is* the OOB check.

File: imsApp/src/imsRecord.cc
Comment (mcb64) [reply imsApp/src/imsRecord.cc:3535]: For reasons that escape me, EPICS removed the pthread lock parameter from their semaphore (epicsEvent).  If this was just pthread code, we'd be fine without the extra check, as they have an explicit lock that is released as you wait and relocked when you timeout.

File: sim/Makefile
Comment (mcb64) [reply sim/Makefile:7]: This is generally not compiled at all.  I wanted to test, and I knew it would take ages for me to get a proper IMS test setup, so I wrote a simulator.  I only included it here because it was too potentially useful to just delete.

File: imsApp/src/imsRecord.cc
Comment (mcb64) [reply imsApp/src/imsRecord.cc:3530]: This goes back to @joshc-slac's comment about my paranoia, above.  The pthread library has a lock parameter for their semaphore wait... the *library* unlocks while waiting, and relocks when returning.

Sigh.

File: imsApp/Db/ims_long.db
Comment (mcb64) [reply imsApp/Db/ims_long.db:706]: (Oh, and I should mention... no Beckhoff here.  IMS is its own style of controller.)

File: imsApp/Db/ims_short.db
Comment (mcb64) [reply imsApp/Db/ims_short.db:650]: Same value, different format.  I don't think it's worthwhile pointing out the format difference.

File: imsApp/src/imsRecord.cc
Comment (mcb64) [reply imsApp/src/imsRecord.cc]: Restored in next checkin.

File: imsApp/src/imsRecord.cc
Comment (mcb64) [reply imsApp/src/imsRecord.cc:169]: Changing the EPICS world to c11 is unlikely to happen during my lifetime.

File: imsApp/src/imsRecord.cc
Comment (mcb64) [reply imsApp/src/imsRecord.cc:3660]: While I agree that this is an overly long file, I don't think that this is really very shareable.  This is very IMS-motor specific... and more than that, it is very specific to IMS motors running *our* PCDS Mcode.

File: imsApp/src/imsRecord.cc
Comment (mcb64) [reply imsApp/src/imsRecord.cc:3670]: For such a small number of possible write requests, I don't mind just looping.  If we wanted to be more efficient, yes, we could actually queue the requests, and then just see if there is anything in the queue here.

We can't get notifications about the read though.  Interestingly enough, while the AsynOctet framework does provide a callback when there is input to be read, this is not actually implemented in asyn.  (I tried... first pass was to use the asyn read callback, and I didn't get anything.  Then I found this: https://epics.anl.gov/tech-talk/2019/msg00617.php.  Checking the most recent asyn, this has still not been implemented.)

Since we have to sleep on the read, a quick poll here should work fine.

File: imsApp/src/imsRecord.cc
Comment (mcb64) [reply imsApp/src/imsRecord.cc]: OK, I actually tested this... even if the server (IMS motor) sends two messages at once, these will be received by asyn as a single message, but passed to the read call as two separate messages.

I've removed this in a forthcoming checkin.

File: imsApp/src/imsRecord.cc
Comment (mcb64) [reply imsApp/src/imsRecord.cc]: This has been removed.  See previous.

File: imsApp/src/imsRecord.cc
Comment (mcb64) [reply imsApp/src/imsRecord.cc:3537]: Yeah, brackets added.  And thinking of it, I did this to the pass test above as well.

File: imsApp/src/imsRecord.cc
Comment (mcb64) [reply imsApp/src/imsRecord.cc]: I've never really been a fan of auto.  But I suppose that's just because I'm old. :-)

File: imsApp/src/imsRecord.cc
Comment (joshc-slac) [reply imsApp/src/imsRecord.cc:157]: Yes, I am talking about the 3 indexes into the array of `ims_asyn_req`s below. 

I see here it is effectively defined as safe in the ims.h file. 
https://github.com/slac-epics/IMS/blob/5d3615fd2472b8c70cbf17e2deb008b9e0c93d4b/imsApp/src/ims.h#L25-L26

Still good practice to check the size of an array before indexing into it; effectively an in situ unit test for future devs.

File: imsApp/src/imsRecord.cc
Comment (joshc-slac) [reply imsApp/src/imsRecord.cc]: Understood. I see it as one line of code to explicitly future proof. Safe as is, just calling out code smell

**Review by joshc-slac — APPROVED** [review]
LGTM: would like to know if this has been tested in our production environment. 

I know general c11 style guide things can come across as nitpicky so thank you for playing ball. This is a larger PR with its whole own feature set, we can consider any robustification work as out of scope ; I will be more firm with future PRs, every time we touch this we should try make it conform more to the past decade+ of industry standards it will make our codebase more extensible, debugable, and lower the barrier to access. That being said, a lot of this code hasn't been touched in 1-10 years, any refactoring would also require unit testing to increase test coverage and ensure we don't enter refactoring hell.

**Comment by mcb64** [comment]
> LGTM: would like to know if this has been tested in our production environment.

Not yet.  Still waiting for a test setup.  (This is why I wrote the simulator in the first place.)

I have no plans to merge/release this until I can test it with an actual motor.

**Review by slactjohnson — COMMENTED** [review]
I worked through all the changes, and LGTM. I didn't see any issues. 
The simulator is a nice utility but, as noted, testing on a real motor is ideal. Will move to "approved" when IRL testing looks good.
I haven't used the "Smart" motors in quite a while, but I know there are some that are running special/different firmware. Any concerns of incompatibility there?

**Comment by mcb64** [comment]
Sigh.  Nothing is ever as simple as you would expect.  If you send the message:

- PR "PN=",PN
You might expect a response like:

- PN=MDI3CRL23C7-EQ
However, what you get is:

- MDI3CRL23C7-EQPN=

This last checkin looks for the "expected" characters either at the beginning or end of the string.

**Comment by mcb64** [comment]
> I haven't used the "Smart" motors in quite a while, but I know there are some that are running special/different firmware. Any concerns of incompatibility there?

It looks like the print bug I mentioned in my last checkin is one of these.  I noticed that there was code looking at the beginning or end of the response string already... so I assume that some motors have this issue, and some don't.

Fortunately, I tested on a "have".

**Pull request merged**
